chapter/1/rpc-reviewed-dont-keep.md

```diff
@@ -0,0 +1,619 @@
+microservices system.
+
+An example of a microservices ecosystem that uses futures/promises is Finagle at
+Twitter.
+
+## Security in RPC:
+
+The initial RPC implementation {% cite implementingrpc --file rpc %} was
+developed for an isolated LAN network and didn't focus much on security.
+There're various attack surfaces in that model, from the malicious registry, to
+a malicious server, to a client targeting for Denial-of-Service to
+Man-in-the-Middle attack between client and server.
+
+As time progressed and internet evolved, new standards came along, and RPC
+implementations became much more secure. Security, in RPC, is generally added as
+a *module* or a *package*. These modules have libraries for authentication and
+authorization of the communication services (caller and callee). These modules
+are not always bug-free and it's possible to gain unauthorized access to the
+system. Efforts are being made to rectify these situations by the security in
+general, using code inspection and bug bounty programs to catch these bugs
+before-hand. However, with time new bugs arise and this cycle continues. It's a
+vicious cycle between attackers and security experts, both of whom tries to
+outdo their opponent.
+
+For example, the Oracle Network File System uses a *Secure RPC*{% cite oraclenfs
+--file rpc %} to perform authentication in the NFS. This *Secure RPC* uses
+Diffie-Hellman authentication mechanism with DES encryption to allow only
+authorized users to access the NFS. Similarly, Cap'n Proto {% cite
+capnprotosecure --file rpc %} claims that it is resilient to memory leaks,
+segfaults, and malicious inputs and can be used between mutually untrusting
+parties. However, in Cap'n Proto "the RPC layer is not robust against resource
+exhaustion attacks, possibly allowing denials of service", nor has it undergone
+any formal verification {% cite capnprotosecure --file rpc %}.
+
+Although, it's possible to come up with a *Threat Model* that would make an RPC
+implementation insecure to use, however, one has to understand that using any
+distributed system increases the attack surface anyways and claiming one
+*paradigm* to be more secure than another would be a biased statement, since
+*paradigms* are generally an idea and it depends on different system designers
+to use these *paradigms* to build their systems and take care of features
+specific to real systems, like security and load-balancing. There's always a
+possibility of rerouting a request to a malicious server(if the registry gets
+hacked), or there's no trust between the *caller* and *callee*. However, we
+maintain that RPC *paradigm* is not secure or insecure(for that matter), and
+that the most secure systems are the ones that are in an isolated environment,
+disconnected from the public internet with a self-destruct mechanism{% cite self
+--file rpc %} in place, in an impenetrable bunker, and guarded by the Knights
+Templar(*they don't exist! Well, maybe Fort Meade comes close*).
```

Comment:
You might laugh, but that is exactly how the Pentagon operates their nuclear weapons :)  They are only accessible via floppy-disks, with no network connection at all - below is the link to it:

http://time.com/4348494/pentagon-nuclear-floppy-disks/

~p